_posts/2019-05-29-canary-releases.md

```diff
@@ -0,0 +1,189 @@
+
+## Background
+
+Our teams strive to keep a high level of urgency for delivery. Therefore they maintain their delivery pipelines themselves. An inquiry across our continuous delivery (CD) pipelines showed that our teams use one of the following designs:
+
+- *Production follows Staging*: This is the most classic design among all. It prevails in services with older staged workflows where changes are tested in an isolated staging environment without real user traffic.
+- *Production with Feature Flags*: This workflow is in place for a constant high pace of changes, especially with UX impact.
+- *Production with A/B Tests*: Another variation of the last design is to keep multiple versions of the system online (e.g., A and B version) and split user traffic manually by an operator.
+
+Although all three designs have a positive impact on our release quality already, they are still very tedious to operate or widen the human error vector. To minimize toil and human errors, we introduced another complementary release strategy — canary releases.
+
+## Canary Releases: Our Design
+
+In short canary releases is an automation extension for our CD pipelines to compare a new release (*the canary group*) against the previous version. Ideally, the old deployment (*the main group*) is not touched by this operation. Instead, a new deployment with the old configuration, *the control group*, is created at the same time as the canary.
+
+Our design is based upon a strict set of decisions:
+
+1. User traffic needs to be split across the main deployment and the other two groups, whereas canary and control need an equal traffic share to keep comparisons sane.
+2. The CD pipeline needs a data source (e.g., metrics, logs, etc.) to evaluate the canary soundness in comparison to the control instance. The decisions can vary from shifting more traffic to the canary/control group, take canary down or replace the current main with the canary.
+3. The three instance groups need to operate independently from each other in isolation.
+
+## Enter Traefik Splitting
+
+One of the significant benefits to using Traefik is that we can rely on building on low-entry barrier features. Although our platform is hosted on GKE, we still need to tailor features according to our use cases. Canary releases being one of them requires us to split traffic across deployments.
+
+Traefik being our single proxy to route traffic to our deployments, has a [built-in feature](https://docs.traefik.io/user-guide/kubernetes/#traffic-splitting) to split traffic across deployment groups through a single Ingress. Therefore a canary deployment can be accomplished with the following Ingress specification:
+
+```
```

Comment:
Tip: you can also, if you want to, add this tag with info, what the snippet is about, e.g.:

```html
<em class="snippet-description">Configuration example for ... bla bla bla</em>
```